File: cgsv/mrs_2.0.0/mrs/sample/cpp/source/echo_client.cpp
#include <mrs.hpp>
#include <mrs/connect.hpp>

static bool         g_IsRun            = true;
static bool         g_IsKeyExchange    = true;
static bool         g_IsEncryptRecords = true;
static uint16       g_RecordOptions    = 0;
static uint32       g_WriteDataLen     = 0;
static uint32       g_WriteCount       = 0;
static uint32       g_Connections      = 0;
static uint32       g_ReadCount        = 0;
static char         g_Header[ 128 ];
static mrs::Connect g_Connect;
static bool         g_IsValidRecord    = true;
static const char*  g_ConnectionPath   = "";

std::string connection_type_to_string( MrsConnection connection ){
    MrsConnectionType type = mrs_connection_get_type( connection );
    switch ( type ){
    case MRS_CONNECTION_TYPE_NONE:{ return "NONE"; }
    case MRS_CONNECTION_TYPE_TCP:{ return "TCP"; }
    case MRS_CONNECTION_TYPE_UDP:{ return "UDP"; }
    case MRS_CONNECTION_TYPE_WS:{ return "WS"; }
    case MRS_CONNECTION_TYPE_WSS:{ return "WSS"; }
    case MRS_CONNECTION_TYPE_TCP_SSL:{ return "TCP_SSL"; }
    case MRS_CONNECTION_TYPE_MRU:{ return "MRU"; }
    }
    return "INVALID";
}

void read_echo( MrsConnection connection, const void* payload, uint32 payload_len ){
    mrs::Time read_time;
    read_time.Set();
    mrs::Buffer buffer;
    buffer.Write( payload, payload_len );
    while ( 0 < buffer.GetDataLen() ){
        mrs::Time write_time = buffer.ReadTime();
        MRS_LOG_DEBUG( "read_echo data=%s data_len=%u diff_time=%s(%s - %s)",
            buffer.GetData(), g_WriteDataLen,
            ( read_time - write_time ).ToString().c_str(), read_time.ToString().c_str(), write_time.ToString().c_str() );
        if ( ! buffer.Read( NULL, g_WriteDataLen ) ){
            MRS_LOG_ERR( "Lost data. len=%u %s", buffer.GetDataLen(), mrs::Utility::ToHex( buffer.GetData(), buffer.GetDataLen() ).c_str() );
            g_IsRun = false;
            break;
        }
        
        ++g_ReadCount;
        if ( g_WriteCount * g_Connections <= g_ReadCount ){
            MRS_LOG_DEBUG( "Since all records have been received, it is finished." );
            g_IsRun = false;
        }
    }
}

void parse_record( MrsConnection connection, void* connection_data, uint32 seqnum, uint16 options, uint16 payload_type, const void* payload, uint32 payload_len ){
    MRS_LOG_DEBUG( "parse_record seqnum=%u options=0x%02X payload=0x%02X/%u", seqnum, options, payload_type, payload_len );
    // MRS_PAYLOAD_TYPE_BEGIN - MRS_PAYLOAD_TYPE_ENDの範囲内で任意のIDを定義し、対応するアプリケーションコードを記述する
    switch ( payload_type ){
    case 0x01:{
        read_echo( connection, payload, payload_len );
    }break;
    
    default:{}break;
    }
}

void write_echo( MrsConnection connection, char* data, uint32 data_len ){
    mrs::Time write_time;
    write_time.Set();
    mrs::Buffer buffer;
    buffer.WriteTime( &write_time );
    buffer.Write( data, data_len );
    
    if ( g_IsValidRecord ){
        mrs_write_record( connection, g_RecordOptions, 0x01, buffer.GetData(), buffer.GetDataLen() );
    }else{
        mrs_write( connection, buffer.GetData(), buffer.GetDataLen() );
    }
}

void write_echo_all( MrsConnection connection ){
    uint32 header_len = snprintf( g_Header, sizeof( g_Header ), "%s ", connection_type_to_string( connection ).c_str() );
    header_len += snprintf( &g_Header[ header_len ], sizeof( g_Header ) - header_len, "%s", ( g_IsKeyExchange && g_IsEncryptRecords ) ? "  CRYPT" : "NOCRYPT" );
    
    char* data = (char*)MRS_MALLOC( g_WriteDataLen );
    if ( NULL == data ){
        MRS_LOG_ERR( "write_echo_all malloc error: size=%u", g_WriteDataLen );
        return;
    }
    
    memset( data, 0xFF, g_WriteDataLen );
    for ( uint32 i = 0; i < g_WriteCount; ++i ){
        snprintf( data, g_WriteDataLen, "%s %llu: %u", g_Header, (uint64)connection, i + 1 );
        write_echo( connection, data, g_WriteDataLen );
    }
    MRS_FREE( data );
}

// 鍵交換した時に呼ばれる
void on_key_exchange( MrsConnection connection, void* connection_data ){
    MRS_LOG_DEBUG( "on_key_exchange" );
    
    write_echo_all( connection );
}

// ソケット接続時に呼ばれる
void on_connect( MrsConnection connection, void* connection_data ){
    MRS_LOG_DEBUG( "on_connect local_mrs_version=0x%x remote_mrs_version=0x%x",
        mrs_get_version( MRS_VERSION_KEY ), mrs_connection_get_remote_version( connection, MRS_VERSION_KEY ) );
    
    if ( g_IsKeyExchange ){
        mrs_set_cipher( connection, mrs_cipher_create( MRS_CIPHER_TYPE_ECDH ) );
        mrs_key_exchange( connection, on_key_exchange );
    }else{
        write_echo_all( connection );
    }
}

// ソケット切断時に呼ばれる
void on_disconnect( MrsConnection connection, void* connection_data ){
    MRS_LOG_DEBUG( "on_disconnect local_mrs_version=0x%x remote_mrs_version=0x%x",
        mrs_get_version( MRS_VERSION_KEY ), mrs_connection_get_remote_version( connection, MRS_VERSION_KEY ) );
}

// ソケットにエラーが発生した時に呼ばれる
void on_error( MrsConnection connection, void* connection_data, MrsConnectionError status ){
    switch ( status ){
    case MRS_CONNECT_ERROR:
    case MRS_CONNECT_TIMEOUT:{
        MrsConnection client = g_Connect.FallbackConnect( connection );
        if ( NULL != client ) return;
    }break;
    
    default: break;
    }
    
    MRS_LOG_DEBUG( "on_error local_mrs_version=0x%x remote_mrs_version=0x%x status=%s",
        mrs_get_version( MRS_VERSION_KEY ), mrs_connection_get_remote_version( connection, MRS_VERSION_KEY ), mrs_get_connection_error_string( status ) );
}

// レコード受信時に呼ばれる
void on_read_record( MrsConnection connection, void* connection_data, uint32 seqnum, uint16 options, uint16 payload_type, const void* payload, uint32 payload_len ){
    parse_record( connection, connection_data, seqnum, options, payload_type, payload, payload_len );
}

// バイナリデータ受信時に呼ばれる
void on_read( MrsConnection connection, void* connection_data, const void* data, uint32 data_len ){
    read_echo( connection, data, data_len );
}

// フォールバック接続時に呼ばれる
void on_fallback_connect( MrsConnection connection, const mrs::Connect::Request* request ){
    MRS_LOG_DEBUG( "on_fallback_connect connection_type=%d addr=%s port=%u timeout_msec=%u", (int)request->connection_type, request->addr, request->port, request->timeout_msec );
    mrs_set_connect_callback( connection, on_connect );
    mrs_set_disconnect_callback( connection, on_disconnect );
    mrs_set_error_callback( connection, on_error );
    
    if ( g_IsValidRecord ){
        mrs_set_read_record_callback( connection, on_read_record );
    }else{
        mrs_set_read_callback( connection, on_read );
    }
    mrs_connection_set_path( connection, g_ConnectionPath );
}

int main( int argc, char** argv ){
    int argi = 1;
    MRS_LOG_DEBUG( "arg %02d: connection_type(TCP:%d UDP:%d WS:%d WSS:%d TCP_SSL:%d MRU:%d TCP -> WSS -> WS:*) [1]", argi++, MRS_CONNECTION_TYPE_TCP, MRS_CONNECTION_TYPE_UDP, MRS_CONNECTION_TYPE_WS, MRS_CONNECTION_TYPE_WSS, MRS_CONNECTION_TYPE_TCP_SSL, MRS_CONNECTION_TYPE_MRU );
    MRS_LOG_DEBUG( "arg %02d: is_key_exchange(OFF:0 ON:1) [1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: is_encrypt_records(OFF:0 ON:1) [1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: write_data_len [1024]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: write_count [10]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: connections [1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: sleep_msec [1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: server_addr [127.0.0.1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: server_port [22222]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: timeout_msec [5000]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: is_valid_record(OFF:0 ON:1) [1]", argi++ );
    MRS_LOG_DEBUG( "arg %02d: connection_path [/]", argi++ );
    
    argi = 1;
    const char* arg_connection_type    = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_is_key_exchange    = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_is_encrypt_records = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_write_data_len     = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1024";
    const char* arg_write_count        = ( ++argi <= argc ) ? argv[ argi - 1 ] : "10";
    const char* arg_connections        = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_sleep_msec         = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_server_addr        = ( ++argi <= argc ) ? argv[ argi - 1 ] : "127.0.0.1";
    const char* arg_server_port        = ( ++argi <= argc ) ? argv[ argi - 1 ] : "22222";
    const char* arg_timeout_msec       = ( ++argi <= argc ) ? argv[ argi - 1 ] : "5000";
    const char* arg_is_valid_record    = ( ++argi <= argc ) ? argv[ argi - 1 ] : "1";
    const char* arg_connection_path    = ( ++argi <= argc ) ? argv[ argi - 1 ] : "/";
    MRS_LOG_DEBUG( "connection_type=%s is_key_exchange=%s is_encrypt_records=%s write_data_len=%s write_count=%s connections=%s sleep_msec=%s server_addr=%s server_port=%s timeout_msec=%s is_valid_record=%s connection_path=%s",
        arg_connection_type, arg_is_key_exchange, arg_is_encrypt_records, arg_write_data_len, arg_write_count, arg_connections,
        arg_sleep_msec, arg_server_addr, arg_server_port, arg_timeout_msec,
        arg_is_valid_record, arg_connection_path );
    
    g_IsKeyExchange = ( 0 != atoi( arg_is_key_exchange ) );
    g_IsEncryptRecords = ( 0 != atoi( arg_is_encrypt_records ) );
    g_RecordOptions = g_IsEncryptRecords ? MRS_RECORD_OPTION_ON_CRYPT : MRS_RECORD_OPTION_NONE;
//    g_RecordOptions |= MRS_RECORD_OPTION_UDP_UNRELIABLE;
//    g_RecordOptions |= MRS_RECORD_OPTION_UDP_UNSEQUENCED;
    g_WriteDataLen = strtoul( arg_write_data_len, NULL, 0 );
    g_WriteCount = strtoul( arg_write_count, NULL, 0 );
    g_Connections = strtoul( arg_connections, NULL, 0 );
    g_IsValidRecord = ( 0 != atoi( arg_is_valid_record ) );
    g_ConnectionPath = arg_connection_path;
    
    uint32 sleep_msec = strtoul( arg_sleep_msec, NULL, 0 );
    mrs::Connect::Request connect_request;
    connect_request.connection_type = MRS_CONNECTION_TYPE_NONE;
    connect_request.addr            = arg_server_addr;
    connect_request.port            = (uint16)strtoul( arg_server_port, NULL, 0 );
    connect_request.timeout_msec    = (uint32)strtoul( arg_timeout_msec, NULL, 0 );
    int32 connection_type = atoi( arg_connection_type );
    switch ( connection_type ){
    case MRS_CONNECTION_TYPE_TCP:
    case MRS_CONNECTION_TYPE_UDP:
    case MRS_CONNECTION_TYPE_WS:
    case MRS_CONNECTION_TYPE_WSS:
    case MRS_CONNECTION_TYPE_TCP_SSL:
    case MRS_CONNECTION_TYPE_MRU:{
        connect_request.connection_type = (MrsConnectionType)connection_type;
        g_Connect.AddRequest( &connect_request );
    }break;
    
    default:{
        connect_request.connection_type = MRS_CONNECTION_TYPE_TCP;
        g_Connect.AddRequest( &connect_request );
        connect_request.connection_type = MRS_CONNECTION_TYPE_WSS;
        connect_request.port += 2;
        g_Connect.AddRequest( &connect_request );
        connect_request.connection_type = MRS_CONNECTION_TYPE_WS;
        connect_request.port -= 1;
        g_Connect.AddRequest( &connect_request );
    }break;
    }
    g_Connect.SetFallbackConnectCallback( on_fallback_connect );
    mrs_initialize();
    do{
        uint32 connections = 0;
        for ( uint32 i = 0; i < g_Connections; ++i ){
            MrsConnection client = g_Connect.FallbackConnect();
            if ( NULL == client ){
                MRS_LOG_ERR( "mrs_connect[%u]: %s", i, mrs_get_error_string( mrs_get_last_error() ) );
                break;
            }
            
            ++connections;
        }
        if ( connections != g_Connections ) break;
        
        while ( g_IsRun ){
            mrs_update();
            
            mrs_sleep( sleep_msec );
        }
    }while ( false );
    mrs_finalize();
    return 0;
}

File: cgsv/mrs_2.0.0/mrs/sample/csharp/source/echo_client.cs
using System;
using System.Text;
using System.Threading;
using System.Runtime.InteropServices;
using MrsLibs.Common;
using MrsLibs.Signal;
using MrsLibs.Client;
using MrsLibs.Parser;

namespace echo_client
{
	using MrsConnection = IntPtr;

	/// <summary>
	/// エコークライアントクラス
	/// </summary>
	public class EchoClient : Mrs, IMrsAppFoundation, IDisposable
	{
		/// <summary>
		/// 廃棄フラグ
		/// </summary>
		private bool m_bDisposed = false;

		/// <summary>
		/// 
		/// </summary>
		private static UInt16 m_nRecordOptions = 0;

		/// <summary>
		/// スリープ時間
		/// </summary>
		private UInt32 m_SleepMsec = 1;

		/// <summary>
		/// 鍵交換をする
		/// </summary>
		private static bool m_canKeyExchange = true;

		/// <summary>
		/// 暗号化する
		/// </summary>
		private static bool m_canEncryptRecords = true;

		/// <summary>
		/// エコーデータの書き込み長
		/// </summary>
		private static UInt32 m_nWriteDataLen = 1024;

		/// <summary>
		/// 書き込み回数
		/// </summary>
		private static UInt32 m_nWriteCount = 10;

		/// <summary>
		/// 接続数
		/// </summary>
		private static UInt32 m_nConnections = 1;

		/// <summary>
		/// 読み込み数
		/// </summary>
		private static UInt32 m_nReadCount = 0;

        /// <summary>
        /// レコード受信用
        /// </summary>
        private static bool m_canValidRecord = true;

        /// <summary>
        /// コネクションパス
        /// </summary>
        private static string m_ConnectionPath = "/";

        /// <summary>
        /// 接続タイプ
        /// </summary>
        private MrsConnectionType m_connectionType;

		/// <summary>
		/// サーバーアドレス
		/// </summary>
		private string m_server_addr = "127.0.0.1";

		/// <summary>
		/// サーバーのポート番号
		/// </summary>
		private ushort m_server_port = 22222;

		/// <summary>
		/// タイムアウト値
		/// </summary>
		private UInt32 m_timeout_msec = 5000;

		/// <summary>
		/// ループフラグ
		/// </summary>
		private static volatile bool m_bLoop = true;

		/// <summary>
		/// 接続
		/// </summary>
		private static mrs.Connect g_Connect = new mrs.Connect();

		private static MrsConnectCallback m_OnConnect = OnConnect;
		private static MrsKeyExchangeCallback m_OnKeyExchange = OnKeyExchange;
		private static MrsDisconnectCallback m_OnDisconnect = OnDisconnect;
		private static MrsErrorCallback m_OnError = OnError;
		private static MrsReadRecordCallback m_OnReadRecord = OnReadRecord;
        private static MrsReadCallback m_OnRead = OnRead;
		private static mrs.Connect.FallbackConnectCallback m_FallbackConnectCallback = OnFallbackConnectCallback;

		/// <summary>
		/// 初期化
		/// </summary>
		/// <param name="args">コマンドライン</param>
		public void Initialize(string[] args)
		{
			MrsCmdParser parser = new MrsCmdParser('-');
			parser.Parse(args);

			m_connectionType = (MrsConnectionType)parser.GetNumeric<Int32>("connectionType", (Int32)MrsConnectionType.TCP);
			m_canKeyExchange = (parser.GetNumeric<Int32>("is_key_exchange", 1) != 0) ? true : false;
			m_canEncryptRecords = (parser.GetNumeric<Int32>("is_encrypt_records", 1) != 0) ? true : false;
            m_nRecordOptions = (ushort)((m_canEncryptRecords == true) ? MrsRecordOption.ON_CRYPT : MrsRecordOption.NONE);
            m_nWriteDataLen = parser.GetNumeric<UInt32>("write_data_len", 1024);
			m_nWriteCount = parser.GetNumeric<UInt32>("write_count", 10);
            m_nConnections = parser.GetNumeric<UInt32>("connections", 1);
            m_SleepMsec = parser.GetNumeric<UInt32>("sleep_msec", 1);

            m_server_addr = parser.GetString("server_addr", "127.0.0.1");
			m_server_port = parser.GetNumeric<ushort>("server_port", 22222);
			m_timeout_msec = parser.GetNumeric<UInt32>("timeout_msec", 5000);
            m_canValidRecord = (parser.GetNumeric<Int32>("is_valid_record", 1) != 0) ? true : false;
            m_ConnectionPath = parser.GetString("connection_path", "/");
			
			mrs.Connect.Request connect_request = new mrs.Connect.Request();
			connect_request.ConnectionType = MrsConnectionType.NONE;
			connect_request.Addr = m_server_addr;
			connect_request.Port = m_server_port;
			connect_request.TimeoutMsec = m_timeout_msec;

			switch (m_connectionType)
			{
			case MrsConnectionType.TCP:
			case MrsConnectionType.UDP:
			case MrsConnectionType.WS:
			case MrsConnectionType.WSS:
            case MrsConnectionType.TCP_SSL:
            case MrsConnectionType.MRU:
				{
					connect_request.ConnectionType = m_connectionType;
					g_Connect.AddRequest(connect_request);
				}
				break;

			default:
				{
					connect_request.ConnectionType = MrsConnectionType.TCP;
					g_Connect.AddRequest(connect_request);

					connect_request.ConnectionType = MrsConnectionType.WSS;
					connect_request.Port += 2;
					g_Connect.AddRequest(connect_request);

					connect_request.ConnectionType = MrsConnectionType.WS;
					connect_request.Port -= 1;
					g_Connect.AddRequest(connect_request);
				}
				break;
			}
			g_Connect.SetFallbackConnectCallback(m_FallbackConnectCallback);
		}

		/// <summary>
		/// 実行
		/// </summary>
		public void Run()
		{
			try
			{
				mrs_initialize();
				using (ExitSignal sig = new ExitSignal())
                {
					sig.SetSignal((obj, e) =>
					{
						m_bLoop = false;
						Thread.Sleep(10);
					});

                    uint connections = 0;
                    for(uint i = 0; i < m_nConnections; ++i)
                    {
                        MrsConnection client = g_Connect.FallbackConnect();
                        if (client == IntPtr.Zero)
                        {
                            MRS_LOG_ERR("mrs_connect[{0}]: {1}", i, mrs_get_error_string(mrs_get_last_error()));
                            break;
                        }
                        ++connections;
                    }
                    if (connections != m_nConnections) return;

                    while(m_bLoop)
                    {
                        mrs_update();
                        mrs_sleep(m_SleepMsec);
                    }
                }
			}
			catch (Exception e)
			{
				MRS_LOG_ERR(e.Message);
			}
			finally
			{
				mrs_finalize();
			}
		}

		/// <summary>
		/// 廃棄処理
		/// </summary>
		public void Dispose()
		{
			Dispose(true);
			GC.SuppressFinalize(this);
		}

		/// <summary>
		/// 廃棄処理の実装部
		/// </summary>
		/// <param name="disposed">廃棄フラグ</param>
		public void Dispose(bool disposed)
		{
			if (m_bDisposed == true) return;
			if (m_bDisposed != disposed)
			{
			}
			m_bDisposed = disposed;
		}

		/// <summary>
		/// フォールバック接続時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="request"></param>
		private static void OnFallbackConnectCallback(MrsConnection connection, mrs.Connect.Request request)
		{
			MRS_LOG_DEBUG( "OnFallbackConnectCallback connection_type={0} addr={1} port={2} timeout_msec={3}",
				(int)request.ConnectionType, request.Addr, request.Port, request.TimeoutMsec);

			mrs_set_connect_callback(connection, m_OnConnect);
			mrs_set_disconnect_callback(connection, m_OnDisconnect);
			mrs_set_error_callback(connection, m_OnError);
            if (m_canValidRecord)
            {
    			mrs_set_read_record_callback(connection, m_OnReadRecord);
            }
            else
            {
                mrs_set_read_callback(connection, m_OnRead);
            }
            mrs_connection_set_path(connection, m_ConnectionPath);
		}

		/// <summary>
		/// 鍵交換した時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		private static void OnKeyExchange(MrsConnection connection, IntPtr connection_data)
		{
			MRS_LOG_DEBUG("OnKeyExchange");
			WriteEchoAll(connection);
		}

		/// <summary>
		/// ソケット接続時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		private static void OnConnect(MrsConnection connection, IntPtr connection_data)
		{
			MRS_LOG_DEBUG( "OnConnect local_mrs_version=0x{0} remote_mrs_version=0x{1}",
				mrs_get_version(MRS_VERSION_KEY), mrs_connection_get_remote_version(connection, MRS_VERSION_KEY));

			if (m_canKeyExchange)
			{
				mrs_set_cipher(connection, mrs_cipher_create(MrsCipherType.ECDH));
				mrs_key_exchange(connection, m_OnKeyExchange);
			}
			else
			{
				WriteEchoAll(connection);
			}
		}

		/// <summary>
		/// ソケット切断時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		private static void OnDisconnect(MrsConnection connection, IntPtr connection_data)
		{
			MRS_LOG_DEBUG( "OnDisconnect local_mrs_version=0x{0} remote_mrs_version=0x{1}",
				mrs_get_version(MRS_VERSION_KEY), mrs_connection_get_remote_version(connection, MRS_VERSION_KEY));
		}

		/// <summary>
		/// ソケットにエラーが発生した時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		/// <param name="status"></param>
		private static void OnError(MrsConnection connection, IntPtr connection_data, MrsConnectionError status)
		{
			switch (status)
			{
			case MrsConnectionError.CONNECT_ERROR:
			case MrsConnectionError.CONNECT_TIMEOUT:
				{
					MrsConnection client = g_Connect.FallbackConnect(connection);
					if ( client != IntPtr.Zero ) return;
				}
				break;
			default: 
				break;
			}
	
			MRS_LOG_DEBUG( "OnError local_mrs_version=0x{0} remote_mrs_version=0x{1} status={2}",
				mrs_get_version(MRS_VERSION_KEY), mrs_connection_get_remote_version(connection, MRS_VERSION_KEY), mrs_get_connection_error_string(status));
		}

		/// <summary>
		/// レコード受信時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		/// <param name="seqnum"></param>
		/// <param name="options"></param>
		/// <param name="payload_type"></param>
		/// <param name="_payload"></param>
		/// <param name="payload_len"></param>
		private static void OnReadRecord(MrsConnection connection, IntPtr connection_data, UInt32 seqnum, UInt16 options, UInt16 payload_type, IntPtr _payload, UInt32 payload_len)
		{
			ParseRecord(connection, connection_data, seqnum, options, payload_type, _payload, payload_len);
		}

		/// <summary>
		/// バイナリデータ受信時に呼ばれる
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		/// <param name="_data"></param>
		/// <param name="data_len"></param>
        private static void OnRead(MrsConnection connection, IntPtr connection_data, IntPtr _data, UInt32 data_len)
        {
            ReadEcho(connection, _data, data_len);
        }

		/// <summary>
		/// レコードのパース
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="connection_data"></param>
		/// <param name="seqnum"></param>
		/// <param name="options"></param>
		/// <param name="payload_type"></param>
		/// <param name="_payload"></param>
		/// <param name="payload_len"></param>
		private static void ParseRecord(MrsConnection connection, IntPtr connection_data, UInt32 seqnum, UInt16 options, UInt16 payload_type, IntPtr _payload, UInt32 payload_len)
		{
			Mrs.MRS_LOG_DEBUG("ParseRecord seqnum=0x{0} options=0x{1:X2} payload={2:X}/{3}", seqnum, options, payload_type, payload_len);
			// MRS_PAYLOAD_TYPE_BEGIN - MRS_PAYLOAD_TYPE_ENDの範囲内で任意のIDを定義し、対応するアプリケーションコードを記述する
			switch (payload_type)
			{
			case 0x01:
				ReadEcho(connection, _payload, payload_len);
				break;
			default:
				break;
			}
		}

		/// <summary>
		/// エコーデータの読み込み
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="payload"></param>
		/// <param name="payload_len"></param>
		private static void ReadEcho(MrsConnection connection, IntPtr payload, UInt32 payload_len )
		{
			mrs.Time read_tm = new mrs.Time();
			read_tm.Set();

			byte[] read_data = new byte[payload_len];
			Marshal.Copy(payload, read_data, 0, (int)payload_len);

			mrs.Buffer buffer = new mrs.Buffer();
			buffer.Write(read_data);
            while(buffer.GetDataLen() > 0)
            {
			    mrs.Time write_tm = buffer.ReadTime();
			    MRS_LOG_DEBUG("ReadEcho data={0} data_len={1} diff_time={2}({3} - {4})",
                    ToString(buffer.GetData()), m_nWriteDataLen, (read_tm - write_tm).ToString(), read_tm.ToString(), write_tm.ToString());
                if (!buffer.Read(null, m_nWriteDataLen))
                {
                    MRS_LOG_ERR("Lost data. len={0} {1}", buffer.GetDataLen(), mrs.Utility.ToHex(buffer.GetData(), buffer.GetDataLen()));
                    m_bLoop = false;
                    break;
                }

			    ++m_nReadCount;
			    if (m_nWriteCount * m_nConnections <= m_nReadCount)
			    {
				    MRS_LOG_DEBUG("Since all records have been received, it is finished.");
				    m_bLoop = false;
			    }
            }
		}

		/// <summary>
		/// エコーデータを書き込み数分書き込む
		/// </summary>
		/// <param name="connection"></param>
		private static void WriteEchoAll(MrsConnection connection)
		{
			StringBuilder sb = new StringBuilder();
			switch (g_Connect.GetRequest().ConnectionType)
			{
			case MrsConnectionType.TCP:
				sb.Append("TCP ");
				break;
			case MrsConnectionType.UDP:
				sb.Append("UDP ");
				break;
			case MrsConnectionType.WS:
				sb.Append("WS ");
				break;
			case MrsConnectionType.WSS:
				sb.Append("WSS ");
				break;
            case MrsConnectionType.TCP_SSL:
                sb.Append("TCP_SSL");
                break;
            case MrsConnectionType.MRU:
                sb.Append("MRU");
                break;
			default:
                sb.Append("INVALID");
				break;
			}

			if (m_canKeyExchange && m_canEncryptRecords) 
				sb.Append("  CRYPT");
			else
				sb.Append("NOCRYPT");

			byte[] data = new byte[m_nWriteDataLen];
			for(uint i = 0; i < m_nWriteCount; ++i)
			{
				string str = String.Format("{0} {1}: {2}", sb.ToString(), connection, i + 1);
				byte[] conv = System.Text.Encoding.UTF8.GetBytes(str);
				System.Buffer.BlockCopy(conv, 0, data, 0, conv.Length);
				WriteEcho(connection, data, m_nWriteDataLen);
			}
		}

		/// <summary>
		/// エコーデータの書き込み
		/// </summary>
		/// <param name="connection"></param>
		/// <param name="data"></param>
		private static void WriteEcho(MrsConnection connection, byte[] data, uint data_len)
		{
			mrs.Time write_tm = new mrs.Time();
			write_tm.Set();
			mrs.Buffer buffer = new mrs.Buffer();
			buffer.WriteTime(write_tm);
			buffer.Write(data, data_len);

            if (m_canValidRecord)
            {
    			mrs_write_record(connection, m_nRecordOptions, 0x01, buffer.GetData(), buffer.GetDataLen());
            }
            else
            {
                mrs_write(connection, buffer.GetData(), buffer.GetDataLen());
            }
		}
	}

	/// <summary>
	/// メインクラス
	/// </summary>
	public class MrsMain
	{
		/// <summary>
		/// メイン関数
		/// </summary>
		/// <param name="args">コマンドライン</param>
		public static void Main(string[] args)
		{
			using (EchoClient client = new EchoClient())
			{
				if (client != null)
				{
					client.Initialize(args);
					client.Run();
				}
			}
		}
	}
}
